keanu-project/src/test/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/operators/unary/MatrixDeterminantVertexTest.java

```diff
@@ -0,0 +1,73 @@
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import io.improbable.keanu.algorithms.variational.optimizer.Optimizer;
+import io.improbable.keanu.network.BayesianNetwork;
+import io.improbable.keanu.tensor.TensorMatchers;
+import io.improbable.keanu.tensor.dbl.DoubleTensor;
+import io.improbable.keanu.vertices.dbl.Differentiator;
+import io.improbable.keanu.vertices.dbl.DoubleVertex;
+import io.improbable.keanu.vertices.dbl.nonprobabilistic.ConstantDoubleVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.GaussianVertex;
+import io.improbable.keanu.vertices.dbl.probabilistic.UniformVertex;
+
+public class MatrixDeterminantVertexTest {
+    @Test
+    public void calculatesDeterminant() {
+        final DoubleVertex input = new ConstantDoubleVertex(DoubleTensor.create(new double[]{1, 2, 3, 4}, 2, 2));
+        Assert.assertThat(input.matrixDeterminant().getValue(), TensorMatchers.isScalarWithValue(-2d));
+    }
+
+    @Test
+    public void canDifferentiateWrtScalar() {
+        final int[] shape = new int[]{2, 2};
+        final DoubleVertex input = new UniformVertex(shape, 0, 10);
+        input.setValue(DoubleTensor.create(new double[]{1, 2, 3, 4}, shape));
+        final DoubleVertex output = input.matrixDeterminant().plus(1);
+        final DoubleTensor expectedDerivative = DoubleTensor.create(new double[]{4, -3, -2, 1}, shape);
+        assertReverseAutoDiffMatches(output, input, expectedDerivative);
+    }
+
+    @Test
+    public void canDifferentiateWrtTensor() {
+        final int[] shape = new int[]{2, 2};
+        final DoubleVertex input = new UniformVertex(shape, 0, 10);
+        input.setValue(DoubleTensor.create(new double[]{1, 2, 3, 4}, shape));
+        final DoubleVertex output = input.matrixDeterminant().times(input);
+
+        // For calculation of derivative, see
+        // https://www.wolframalpha.com/input/?i=x+%3D+1;+y+%3D+2;+z+%3D+3;+w+%3D+4;+d%2Fdx+(%7B%7Bx,y%7D,%7Bz,w%7D%7D+*+det(%7B%7Bx,y%7D,%7Bz,w%7D%7D))
+        final double[] expectedDerivativeValues = new double[]{2, -3, -2, 1, 8, -8, -4, 2, 12, -9, -8, 3, 16, -12, -8, 2};
+        final DoubleTensor expectedDerivative = DoubleTensor.create(expectedDerivativeValues, 2, 2, 2, 2);
+        assertReverseAutoDiffMatches(output, input, expectedDerivative);
+    }
+
+    @Test
+    public void canOptimiseOutOfTheBoxEvenWithRiskOfSingularMatrices() {
+        // The standard method for calculating the gradient of the determinant fails when the matrixDeterminant is 0.
+        // Regardless, the gradient is defined when the matrixDeterminant is 0.
+        // We run the optimizer many times to create some confidence that we won't run into problems in a real-world
+        // use-case.
+        final int runCount = 100;
+        IntStream.range(0, runCount).forEach(i -> {
+            int[] shape = new int[]{2, 2};
+            DoubleVertex input = new GaussianVertex(shape, Math.random() * 2 - 4, Math.random() * 2 + 0.1);
```

Comment:
Nondeterministic tests are really a pain. Think about the edge cases this random test might find and then test them explicitly and deterministically.

Reply:
That's a good point. This loop is also making the test take much longer than it needs to.